src/cargo/util/flock.rs

```diff
@@ -193,57 +195,78 @@ impl Filesystem {
+        opts.read(true).write(true).create(true);
+        let (path, f) = self.open(path.as_ref(), &opts, true)?;
+        self.lock(path, f, config, msg, false)
+    }
 
-        // If we want an exclusive lock then if we fail because of NotFound it's
-        // likely because an intermediate directory didn't exist, so try to
-        // create the directory and then continue.
+    /// A non-blocking version of [`Filesystem::open_shared_create`].
+    ///
+    /// Returns `None` if the operation would block due to another process
+    /// holding the lock.
+    pub fn try_open_shared_create<P: AsRef<Path>>(&self, path: P) -> CargoResult<Option<FileLock>> {
+        let mut opts = OpenOptions::new();
+        opts.read(true).write(true).create(true);
+        let (path, f) = self.open(path.as_ref(), &opts, true)?;
+        if try_acquire(&path, &|| try_lock_shared(&f))? {
+            Ok(Some(FileLock { f: Some(f), path }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn open(&self, path: &Path, opts: &OpenOptions, create: bool) -> CargoResult<(PathBuf, File)> {
+        let path = self.root.join(path);
         let f = opts
             .open(&path)
             .or_else(|e| {
-                if e.kind() == io::ErrorKind::NotFound && state == State::Exclusive {
+                // If we were requested to create this file, and there was a
+                // NotFound error, then that was likely due to missing
+                // intermediate directories. Try creating them and try again.
+                if e.kind() == io::ErrorKind::NotFound && create {
                     paths::create_dir_all(path.parent().unwrap())?;
                     Ok(opts.open(&path)?)
                 } else {
                     Err(anyhow::Error::from(e))
                 }
             })
             .with_context(|| format!("failed to open: {}", path.display()))?;
-        match state {
-            State::Exclusive => {
-                acquire(config, msg, &path, &|| try_lock_exclusive(&f), &|| {
-                    lock_exclusive(&f)
-                })?;
-            }
-            State::Shared => {
-                acquire(config, msg, &path, &|| try_lock_shared(&f), &|| {
-                    lock_shared(&f)
-                })?;
-            }
-            State::Unlocked => {}
+        Ok((path, f))
+    }
+
+    fn lock(
+        &self,
+        path: PathBuf,
+        f: File,
+        config: &Config,
+        msg: &str,
+        exclusive: bool,
+    ) -> CargoResult<FileLock> {
+        if exclusive {
+            acquire(config, msg, &path, &|| try_lock_exclusive(&f), &|| {
+                lock_exclusive(&f)
+            })?;
+        } else {
+            acquire(config, msg, &path, &|| try_lock_shared(&f), &|| {
+                lock_shared(&f)
+            })?;
         }
-        Ok(FileLock {
-            f: Some(f),
-            path,
-            state,
-        })
+        Ok(FileLock { f: Some(f), path })
```

Comment:
Is this still worth it?

I felt this made the code more confusing due to the lack of parallels between the try and non-try variants.

Reply:
I'm not sure if I'm understanding. I assume this is asking if this should just be inlined? That seems fine to me. I don't like bool arguments very much, but I felt like the `acquire` call was complex enough to justify trying to share it.

Reply:
Thanks!